File: native-array-memory/js-2/buffer.js
"use strict";

function Buffer (size, init_value) {
  this.size = size;
  this.array = new Array(size);
  this.array.fill(init_value);
  this.cursor = 0;
};

Buffer.prototype = {
  set: function (index, value) {
    this.array[index] = value;
  },

  get: function (index) {
    return this.array[index];
  },

  add_cursor: function (value) {
    this.cursor = this.cursor + value;
  },

  allocate: function (size) {
    let return_address = this.cursor;
    this.add_cursor(size);
    return return_address;
  }
};

File: native-array-memory/clojure/interpreter/src/interpreter/buffer.clj
(ns interpreter.buffer)

(defmacro buffer-set [buffer index value]
  `(aset ~buffer (+ 1 ~index) ~value))

(defmacro buffer-get [buffer index]
  `(aget ~buffer (+ 1 ~index)))

(defn make-buffer [size init-value]
  (let [buffer (object-array (+ size 1))]
    (aset buffer 0 0)
    (loop [counter 0]
      (when (< counter size)
        ;; (aset buffer (+ 1 counter) init-value)
        (buffer-set buffer counter init-value)
        (recur (+ 1 counter))))
    buffer))

(defn buffer-get-cursor [buffer]
  (aget buffer 0))

(defn buffer-add-cursor [buffer value]
  (aset buffer 0
        (+ (buffer-get-cursor buffer) value)))

(defn buffer-allocate [buffer size]
  (let [return-address (swap! buffer buffer-get-cursor)]
    (buffer-add-cursor buffer size)
    return-address))
